### PR TITLE
feat: Add ability to customize notification handler

### PR DESCRIPTION
**Pull request by ghostbuster91**
add `notify.handler` to setup opts

refers to #1697

### CODE REVIEW

File: doc/nvim-tree-lua.txt
Comment (gegoune) [inline doc/nvim-tree-lua.txt:443]: Do we need that default? Can't it be done "behind scenes" and only used if set to function?

Comment (ghostbuster91) [reply doc/nvim-tree-lua.txt:443]: I tried to do it in a similar way to how other options were done (e.g. window picker). I am happy to change it if you want.

Comment (gegoune) [reply doc/nvim-tree-lua.txt:443]: I think it would make sense to have default option implied, with ability to override it.

Comment (ghostbuster91) [reply doc/nvim-tree-lua.txt]: Done, let me know if this is what you had in mind.

Comment (alex-courtis) [reply doc/nvim-tree-lua.txt]: That's kind of a break from existing - everything is specified in `DEFAULT_OPTS`. It's useful if only for the "default schema" that users can base their configuration on.

That's how my local dev harness works: put all the default options in the init lua setup call.

@gegoune are you OK to back that change out? Happy to revisit later; trimming the default options to a skeleton with no values would be less overwhelming.

File: lua/nvim-tree/notify.lua
Comment (alex-courtis) [inline lua/nvim-tree/notify.lua:3]:
```suggestion
local default_handler = function(msg, level, opts)
```

No camel case please.

File: lua/nvim-tree/notify.lua
Comment (alex-courtis) [inline lua/nvim-tree/notify.lua:9]: Unnecessary, set during setup.

File: lua/nvim-tree/notify.lua
Comment (alex-courtis) [inline lua/nvim-tree/notify.lua:38]:
```suggestion
local create_handler = function()
```

No camel case please.

**Review by alex-courtis — CHANGES_REQUESTED** [review]
Tested ok both commits:
* default handler
* `rcarriga/nvim-notify`
* `notify.handler` :
```lua
local function h(msg, level, opts)
	print(vim.inspect(msg))
	print(vim.inspect(level))
	print(vim.inspect(opts))
end
```

Please:

- [x] Update help
- [ ] Remove camel case and rename
- [ ] Revert DEFAULT_OPTS pending @gegoune

File: lua/nvim-tree/notify.lua
Comment (alex-courtis) [inline lua/nvim-tree/notify.lua:54]: This is so much cleaner - many thanks for tidying.

File: doc/nvim-tree-lua.txt
Comment (gegoune) [reply doc/nvim-tree-lua.txt]: Missed that point. Of course. @ghostbuster91 Please reinstate state before my comment and accept my apologies for causing you extra work.

File: doc/nvim-tree-lua.txt
Comment (gegoune) [reply doc/nvim-tree-lua.txt:1238]: Why are we checking for nvim-notify at all? Wasn't idea behind vim.notify such that it handles notifications send to it per users' handler, if it happens to be nvim-notify so be it. 

Not near my computer so can't check much easily.

Comment (alex-courtis) [reply doc/nvim-tree-lua.txt:1238]: Oh I see! We set handler directly to vim.notify.

Much cleaner. Just have to ensure that the arguments line up.

File: doc/nvim-tree-lua.txt
Comment (ghostbuster91) [reply doc/nvim-tree-lua.txt]: No worries, luckily it was just a small change :)

File: doc/nvim-tree-lua.txt
Comment (ghostbuster91) [reply doc/nvim-tree-lua.txt:1238]: > Why are we checking for nvim-notify at all? Wasn't idea behind vim.notify such that it handles notifications send to it per users' handler, if it happens to be nvim-notify so be it.

Yes, that's correct. I did it like that only to preserve the previous behavior, but I would happy to simplify it. Is there any issue/way of notifying users of such change? 

> Much cleaner. Just have to ensure that the arguments line up.

They do :)

File: lua/nvim-tree/notify.lua
Comment (ghostbuster91) [reply lua/nvim-tree/notify.lua]: Indeed, removed.

**Comment by ghostbuster91** [comment]
- [x] Remove camel case and rename
- [x] Revert DEFAULT_OPTS 
- [x] Update help

Waiting for the confirmation regarding auto detection of nvim-notify

File: lua/nvim-tree/notify.lua
Comment (ghostbuster91) [reply lua/nvim-tree/notify.lua]: I've now recalled why I did it like that, it was because of that comment https://github.com/nvim-tree/nvim-tree.lua/pull/1693#discussion_r1009294846 

Let me know if I should restore it

Comment (alex-courtis) [reply lua/nvim-tree/notify.lua]: > I've now recalled why I did it like that, it was because of that comment [#1693 (comment)](https://github.com/nvim-tree/nvim-tree.lua/pull/1693#discussion_r1009294846)
> 
> So the notify module could be called before `setup` function gets called, and in such a case a `fallback_handler` would be used. Let me know if I should restore it.

Ah yes, thank you for the diligence.

File: doc/nvim-tree-lua.txt
Comment (alex-courtis) [reply doc/nvim-tree-lua.txt:1241]: Working:
`handler = require("notify").notify,`
and
`handler = vim.notify,`

the latter case does not prepend the title, however that's expected.

File: doc/nvim-tree-lua.txt
Comment (alex-courtis) [inline doc/nvim-tree-lua.txt:1228]:
```suggestion
    *nvim-tree.notify.handler*
```

File: doc/nvim-tree-lua.txt
Comment (alex-courtis) [reply doc/nvim-tree-lua.txt:1241]: > Yes, that's correct. I did it like that only to preserve the previous behavior, but I would happy to simplify it. Is there any issue/way of notifying users of such change?

If we were building this today, the user would need to specify `handler = require("notify").notify` however the existing auto-detection behaviour exists and I don't like to break behaviour.

How about:
* `notify.handler` is nil or "default" AND
* `require("notify")` successful

"nvim-notify plugin detected: please set notify.handler = require("notify").notify. See :help nvim-tree.notify"

Options:
1. Preserve existing behaviour, defaulting to nvim-notify
2. Remove nvim-notify fallback, requiring the user to explicitly set `notify.handler`

Your call @gegoune :)

**Review by alex-courtis — CHANGES_REQUESTED** [review]

**Comment by gegoune** [comment]
I am all in for breaking changes if justified, they are not as disruptive as people might think if done properly.

Why are we even checking for nvim-notify? So we can call it with some other parameters/if it't not present to prefix message ourselves? If messages sent change then it barely is a breaking change at all. Am I missing on how that is a breaking change? 

I am all in to setting default handler to `vim.notify` with an option to change it to custom function.

(Is `vim.notify` sender aware? If so users that have a need to change handler per plugin could do it in their handler, we wouldn't need to accommodate it at all. :)

**Comment by ghostbuster91** [comment]
> Why are we even checking for nvim-notify? So we can call it with some other parameters/if it't not present to prefix message ourselves? If messages sent change then it barely is a breaking change at all. Am I missing on how that is a breaking change?

The breaking change I had in mind was regarding the auto-detection of nvim-notify. 
I can imagine that some people relay on that feature but for some reason they don't have the `nvim-notify` attached to `vim.notify` directly. I know that it is probably a very rare case but still. I am fine with whatever decision you make, I just wanted to highlight that.

**Comment by gegoune** [comment]
I agree with the removal. If someone is doing something like that we can very easily assume they know what they are doing and will be able to adapt.

**Comment by alex-courtis** [comment]
How about:
* remove it
* put a breaking change note in the readme
* add a note in the :help about `handler = require("notify").notify`

**Comment by gegoune** [comment]
> add a note in the :help about handler = require("notify").notify

I am not sure about that bit. We should just send stuff to `vim.notify`, which is default mechanism de facto now. Purpose of this PR was to enable users to come up with their own handlers. I will need to review it again, why is it required exactly, as once we just use `vim.notify` users can override that handler in Neovim to their liking.

**Comment by ghostbuster91** [comment]
> I will need to review it again, why is it required exactly, as once we just use vim.notify users can override that handler in Neovim to their liking.

Very much indeed! Because of the small changes in the direction that we did during that PR it was not obvious to spot it, but I think that you are right. Now, once the auto-detection of nvim-notify was removed, we can just simply sent everything to `vim.notify` and it is totally up to the user to configure it to their liking.

**Comment by gegoune** [comment]
Ok, good to know I wasn't imagining things. With that change we wouldn't need a custom handler any more. 

Would you like to amend that, or create new, PR with those changes?

**Comment by ghostbuster91** [comment]
I will create a new PR to keep the things clear and refer to this one.

**Pull request closed without merging**
